tests/ui/lint/non-local-defs/weird-exprs.stderr

```diff
@@ -1,71 +1,111 @@
-warning: non-local `impl` definition, they should be avoided as they go against expectation
+warning: non-local `impl` definition, `impl` blocks should be written at the same level as their item
   --> $DIR/weird-exprs.rs:8:5
    |
 LL |     impl Uto for *mut Test {}
-   |     ^^^^^^^^^^^^^^^^^^^^^^^^^
+   |     ^^^^^---^^^^^---------
+   |          |       |
+   |          |       `*mut Test` is not local
+   |          `Uto` is not local
+   |
+   = note: `impl` may be usable in bounds, etc. from outside the expression, which might e.g. make something constructible that previously wasn't, because it's still on a publicly-visible type
+   = note: an `impl` is never scoped, even when it is nested inside an item, as it may impact type checking outside of that item, which can be the case if neither the trait or the self type are at the same nesting level as the `impl`
+help: move this `impl` block outside of the current constant expression `<unnameable>`
```

Comment:
This is a minor gripe, but you should have one message for when the expression is nameable, and one for when it isn't to avoid this.

Edit: it actually seems like every case we show this help it is `<unnameable>`, so we can just remove that part of the message.

Reply:
Instead of this help, why don't we make it a secondary span on the main diagnostic stating something along the lines of "move the `impl` block outside of this ...":

```

warning: non-local `impl` definition, `impl` blocks should be written at the same level as their item
  --> $DIR/weird-exprs.rs:8:5
   |
LL |/ type A = [u32; {
LL ||     impl Uto for *mut Test {}
   ||     ^^^^^---^^^^^---------
   ||          |       |
   ||          |       `*mut Test` is not local
   ||          `Uto` is not local
LL ||     1
LL || }];
   ||___- move the `impl` block outside of this `type`
   |
   = note: `impl` may be usable in bounds, etc. from outside the expression, which might e.g. make something constructible that previously wasn't, because it's still on a publicly-visible type
   = note: an `impl` is never scoped, even when it is nested inside an item, as it may impact type checking outside of that item, which can be the case if neither the trait or the self type are at the same nesting level as the `impl`
   = note: this lint may become deny-by-default in the edition 2024 and higher, see the tracking issue <https://github.com/rust-lang/rust/issues/120363>
   = note: `#[warn(non_local_definitions)]` on by default
```

Reply:
I like the idea, but for how would you expect it to look like for the maximal case (ie where we have multiple suggestions?), putting everything in the main span looks like this:

```text
warning: non-local `impl` definition, `impl` blocks should be written at the same level as their item
  --> $DIR/exhaustive.rs:58:5
   |
LL | fn main() {
   | --------- move this `impl` block outside of the current function `main`
...
LL |     impl Trait for *mut InsideMain {}
   |     ^^^^^-----^^^^^---------------
   |          |         |    |
   |          |         |    may need to be moved as well
   |          |         `*mut InsideMain` is not local
   |          |         help: remove `*mut ` to make the `impl` local
   |          `Trait` is not local
   |
   = note: `impl` may be usable in bounds, etc. from outside the expression, which might e.g. make something constructible that previously wasn't, because it's still on a publicly-visible type
   = note: an `impl` is never scoped, even when it is nested inside an item, as it may impact type checking outside of that item, which can be the case if neither the trait or the self type are at the same nesting level as the `impl`
```
this looks unreadable, any idea how to improve it?

Reply:
Maybe like this, where the "may need to be moved" spans point to the definition type?
```text
warning: non-local `impl` definition, `impl` blocks should be written at the same level as their item
  --> /home/archie/Projects/RustProjects/rust/tests/ui/lint/non-local-defs/exhaustive.rs:58:5
   |
9  | fn main() {
   | --------- move this `impl` block outside of the current function `main`
...
56 |     struct InsideMain;
   |     ----------------- may need to be moved as well
57 |
58 |     impl Trait for *mut InsideMain {}
   |     ^^^^^-----^^^^^---------------
   |          |         |
   |          |         `*mut InsideMain` is not local
   |          |         help: remove `*mut ` to make the `impl` local
   |          `Trait` is not local
   |
   = note: `impl` may be usable in bounds, etc. from outside the expression, which might e.g. make something constructible that previously wasn't, because it's still on a publicly-visible type
   = note: an `impl` is never scoped, even when it is nested inside an item, as it may impact type checking outside of that item, which can be the case if neither the trait or the self type are at the same nesting level as the `impl`
   = note: this lint may become deny-by-default in the edition 2024 and higher, see the tracking issue <https://github.com/rust-lang/rust/issues/120363>
```

Reply:
But that can quickly get out of hand.
```text
warning: non-local `impl` definition, `impl` blocks should be written at the same level as their item
  --> $DIR/suggest-moving-inner.rs:12:5
   |
LL | fn main() {
   | --------- move this `impl` block outside of the current function `main`
LL |     mod below {
LL |         pub struct Type<T>(T);
   |         ------------------ may need to be moved as well
LL |     }
LL |     struct InsideMain;
   |     ----------------- may need to be moved as well
LL |     trait HasFoo {}
   |     ------------ may need to be moved as well
LL |
LL |     impl<T> Trait<InsideMain> for &Vec<below::Type<(InsideMain, T)>>
   |     ^^^^^^^^-----^^^^^^^^^^^^^^^^^----------------------------------
   |             |                     |
   |             |                     `&'_ Vec<below::Type<(InsideMain, T)>>` is not local
   |             `Trait` is not local
   |
   = note: `impl` may be usable in bounds, etc. from outside the expression, which might e.g. make something constructible that previously wasn't, because it's still on a publicly-visible type
   = note: an `impl` is never scoped, even when it is nested inside an item, as it may impact type checking outside of that item, which can be the case if neither the trait or the self type are at the same nesting level as the `impl`
```

Reply:
Or maybe a mix of both?
```text
warning: non-local `impl` definition, `impl` blocks should be written at the same level as their item
  --> /home/archie/Projects/RustProjects/rust/tests/ui/lint/non-local-defs/suggest-moving-inner.rs:12:5
   |
12 |     impl<T> Trait<InsideMain> for &Vec<below::Type<(InsideMain, T)>>
   |     ^^^^^^^^-----^^^^^^^^^^^^^^^^^----------------------------------
   |             |                     |
   |             |                     `&'_ Vec<below::Type<(InsideMain, T)>>` is not local
   |             `Trait` is not local
   |
   = note: `impl` may be usable in bounds, etc. from outside the expression, which might e.g. make something constructible that previously wasn't, because it's still on a publicly-visible type
   = note: an `impl` is never scoped, even when it is nested inside an item, as it may impact type checking outside of that item, which can be the case if neither the trait or the self type are at the same nesting level as the `impl`
help: move this `impl` block outside of the current function `main`
  --> /home/archie/Projects/RustProjects/rust/tests/ui/lint/non-local-defs/suggest-moving-inner.rs:5:1
   |
5  | fn main() {
   | ^^^^^^^^^
6  |     mod below {
7  |         pub struct Type<T>(T);
   |         ------------------ may need to be moved as well
8  |     }
9  |     struct InsideMain;
   |     ----------------- may need to be moved as well
10 |     trait HasFoo {}
   |     ------------ may need to be moved as well
   = note: this lint may become deny-by-default in the edition 2024 and higher, see the tracking issue <https://github.com/rust-lang/rust/issues/120363>
   = note: `#[warn(non_local_definitions)]` on by default
```

Reply:
Okay after some experiment which you can see above, I pushed a mix between your proposition and my last one (for more complicated cases), I think it strikes a good balance, yours becomes the default in the vast majority of cases. Let me known what you think.